sw/device/bsp/w25q/w25q.c

```diff
@@ -0,0 +1,1247 @@
+
+/**
+ * @brief Static vector used in the erase_and_write function.
+ * 
+ * It is used to store the data of a sector (4k) of the flash. Given the dimensions
+ * of the vector, is not possible to allocate it dinamically as it would cause a stack
+ * or heap overflow.
+*/
+uint8_t sector_data[FLASH_SECTOR_SIZE];
+
+
+/****************************************************************************/
+/**                                                                        **/
+/*                           EXPORTED FUNCTIONS                             */
+/**                                                                        **/
+/****************************************************************************/
+uint8_t w25q128jw_init(spi_host_t spi_host) {
+    /*
+     * Check if memory mapped SPI is enabled. Current version of the bsp
+     * does not support memory mapped SPI.
+    */
+    if (soc_ctrl_peri->USE_SPIMEMIO == 1) {
+        return FLASH_ERROR; // Error
+    }
+
+    // Set the global spi variable to the one passed as argument.
+    spi = spi_host;
+
+    #ifdef USE_SPI_FLASH
+    // Select SPI host as SPI output
+    soc_ctrl_select_spi_host(&soc_ctrl);
+    #endif // USE_SPI_FLASH
+    
+    // Enable SPI host device
+    spi_set_enable(&spi, true);
+
+    // Enable SPI output
+    spi_output_enable(&spi, true);
+
+    // Configure SPI<->Flash connection on CSID 0
+    configure_spi();
+
+    // Set CSID
+    spi_set_csid(&spi, 0);
+
+    // Power up flash
+    flash_power_up();
+
+    // Set QE bit (only FPGA, simulation do not support status registers at all)
+    #ifdef TARGET_PYNQ_Z2
+    if (set_QE_bit() == FLASH_ERROR) return FLASH_ERROR; // Error occurred while setting QE bit
+    #endif // TARGET_PYNQ_Z2
+
+    return FLASH_OK; // Success
+}
+
+uint8_t w25q128jw_read(uint32_t addr, void *data, uint32_t length) {
+    // Sanity checks
+    if (sanity_checks(addr, data, length) == 0) return FLASH_ERROR;
+
+    // Define the status variable
+    w25q_error_codes_t status;
+
+    if (length < RX_DMA_THRESHOLD) {
+        status = w25q128jw_read_quad(addr, data, length);
+        if (status != FLASH_OK) return status;
+    } else {
+        // Wait DMA to be free
+        while(!dma_is_ready());
+        status = w25q128jw_read_quad_dma(addr, data, length);
+        if (status != FLASH_OK) return status;
+    }
+
+    return FLASH_OK;
+}
+
+uint8_t w25q128jw_write(uint32_t addr, void *data, uint32_t length, uint8_t erase_before_write) {
+    // Sanity checks
+    if (sanity_checks(addr, data, length) == 0) return FLASH_ERROR;
+
+    // Define the status variable
+    w25q_error_codes_t status;
+
+    if (erase_before_write == 1) {
+        status = erase_and_write(addr, data, length);
+        if (status != FLASH_OK) return status;
+    } else {
+        // Wait DMA to be free
+        while(!dma_is_ready());
+        status = w25q128jw_write_quad_dma(addr, data, length);
+        if (status != FLASH_OK) return status;
```

Comment:
avoid this branch by initializing status as FLASH_OK at the very begining and just returning status at the end of the function